CelesteNet.Client/Components/CelesteNetPico8Component.cs

```diff
@@ -0,0 +1,496 @@
+using Celeste.Mod.CelesteNet.Client.Entities;
+using Celeste.Mod.CelesteNet.DataTypes;
+using Celeste.Pico8;
+using FMOD;
+using Microsoft.Xna.Framework;
+using Monocle;
+using MonoMod.RuntimeDetour;
+using MonoMod.Utils;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace Celeste.Mod.CelesteNet.Client.Components;
+#nullable enable
+
+
+internal class GhostState
+{
+    internal float spr;
+    internal float x;
+    internal float y;
+    internal bool flipX;
+    internal bool flipY;
+    internal int type;
+    internal int djump;
+    internal List<DataPicoState.HairNode> hair = new();
+    internal int level;
+}
+
+public class CelesteNetPico8Component : CelesteNetGameComponent {
+    
+    public readonly ConcurrentDictionary<uint, PicoGhost> ghosts = new();
+
+    readonly GhostState queuedGhostState = new();
+    public bool alive = false;
+
+    public CelesteNetPico8Component(CelesteNetClientContext context, Game game) : base(context, game) {}
+
+    public override void Initialize() {
+        base.Initialize();
+
+        #pragma warning disable CA2012
+        MainThreadHelper.Schedule(() =>
+        {
+            using (new DetourConfigContext(new DetourConfig(
+                "CelesteNetPico8",
+                int.MinValue
+            )).Use())
+            {
+                On.Celeste.Pico8.Classic.player.ctor += OnPlayerCreate;
+                On.Celeste.Pico8.Classic.player_spawn.ctor += OnPlayerSpawn;
+                On.Celeste.Pico8.Classic.player.update += OnPlayerUpdate;
+                On.Celeste.Pico8.Classic.player_spawn.update += OnPlayerSpawnUpdate;
+                On.Celeste.Pico8.Classic.player.draw += OnPlayerDraw;
+                On.Celeste.Pico8.Classic.kill_player += OnPlayerKill;
+                On.Celeste.Pico8.Classic.player_hair.draw_hair += OnDrawHair;
+                On.Celeste.Pico8.Emulator.Update += OnEmulatorUpdate;
+                On.Celeste.Pico8.Emulator.End += OnEmulatorClose;
+                On.Celeste.Pico8.Emulator.ResetScreen += OnEmulatorReset;
+            }
+        });
+
+        CelesteNetPlayerListComponent.OnGetState += ModifyStateOfPicoPlayers;
+        CelesteNetMainComponent.OnSendState += ModifyStateInPico;
+        var client = Context?.Client;
+        if (client != null) {
+            client.Con.OnDisconnect += WipeGhostsOnDisconnect;
+        }
+    }
+
+    private static void SendState() {
```

Comment:
Does the stuff here in SendState need any null checks? And why is it static, if it wasn't you could make use of `CelesteNetGameComponent`'s property `Context` at least